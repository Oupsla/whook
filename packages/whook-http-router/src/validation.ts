import camelCase from 'camel-case';
import YError from 'yerror';
import HTTPError from 'yhttperror';
import Stream from 'stream';
import {
  pickupOperationSecuritySchemes,
  SupportedSecurityScheme,
} from './utils';
import { OpenAPIV3 } from 'openapi-types';
import ajv, { Ajv } from 'ajv';
import { WhookOperation } from '@whook/http-transaction';

/* Architecture Note #1.1: Validators
For performance reasons, the validators are
 created once for all at startup from the
 API definition.

One could argue that it would have been
 better for performances to generate
 the code statically. This is true. It
 may be done later but it won't change
 the way it works so, moving fast for
 now but keeping it in mind.

Also, looking closely to Prepack that
 could improve significantly this
 project performances with close to no
 time costs:
 https://github.com/facebook/prepack/issues/522#issuecomment-300706099
*/

export function applyValidators(
  operation: WhookOperation,
  validators: { [name: string]: ajv.ValidateFunction },
  parameters: any[],
) {
  ((operation.parameters || []) as OpenAPIV3.ParameterObject[]).forEach(
    ({ name, in: isIn }) => {
      if ('header' === isIn) {
        return validators[name](parameters[camelCase(name)]);
      }
      return validators[name](parameters[name]);
    },
  );
}

export function prepareBodyValidator(ajv, operation) {
  if (!(operation.requestBody && operation.requestBody.content)) {
    return _rejectAnyRequestBody;
  }

  const validators = Object.keys(operation.requestBody.content).reduce(
    (validators, mediaType) => {
      const mediaTypeObject = operation.requestBody.content[mediaType];
      const hasNoSchema = !mediaTypeObject.schema;

      if (hasNoSchema) {
        return validators;
      }

      const isBinaryContent =
        mediaTypeObject.schema.type === 'string' &&
        mediaTypeObject.schema.format === 'binary';

      if (isBinaryContent) {
        return validators;
      }

      let validator;

      try {
        validator = ajv.compile(mediaTypeObject.schema);
      } catch (err) {
        throw YError.wrap(
          err,
          'E_BAD_BODY_SCHEMA',
          operation.operationId,
          mediaType,
        );
      }

      return {
        ...validators,
        [mediaType]: validator,
      };
    },
    {},
  );

  return _validateRequestBody.bind(null, validators);
}

function _validateRequestBody(
  validators: ajv.ValidateFunction[],
  operation: WhookOperation,
  contentType: string,
  value: any,
) {
  if (
    (operation.requestBody as OpenAPIV3.RequestBodyObject).required &&
    'undefined' === typeof value
  ) {
    throw new HTTPError(
      400,
      'E_REQUIRED_REQUEST_BODY',
      operation.operationId,
      typeof value,
      (value as any) instanceof Stream ? 'Stream' : value,
    );
  }
  // Streamed contents, let it pass
  if (!validators[contentType]) {
    return;
  }
  if ('undefined' !== typeof value && !validators[contentType](value)) {
    throw new HTTPError(
      400,
      'E_BAD_REQUEST_BODY',
      operation.operationId,
      typeof value,
      value instanceof Stream ? 'Stream' : value,
      validators[contentType].errors,
    );
  }
}

function _rejectAnyRequestBody(
  operation: WhookOperation,
  contentType: string,
  value: any,
) {
  if ('undefined' !== typeof value) {
    throw new HTTPError(
      400,
      'E_NO_REQUEST_BODY',
      operation.operationId,
      typeof value,
      value instanceof Stream ? 'Stream' : value,
    );
  }
}

// Supporting only mainstream schemes
// https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml
const SUPPORTED_HTTP_SCHEMES = ['basic', 'bearer', 'digest'];

export function extractOperationSecurityParameters(
  openAPI: OpenAPIV3.Document,
  operation: WhookOperation,
): OpenAPIV3.ParameterObject[] {
  const operationSecuritySchemes = pickupOperationSecuritySchemes(
    openAPI,
    operation,
  );
  const securitySchemes = Object.keys(operationSecuritySchemes).map(
    schemeKey => operationSecuritySchemes[schemeKey],
  );

  return extractParametersFromSecuritySchemes(securitySchemes);
}

export function extractParametersFromSecuritySchemes(
  securitySchemes: (SupportedSecurityScheme | OpenAPIV3.OpenIdSecurityScheme)[],
): OpenAPIV3.ParameterObject[] {
  const hasOAuth = securitySchemes.some(securityScheme =>
    ['oauth2', 'openIdConnect'].includes(securityScheme.type),
  );
  const httpSchemes = [
    ...new Set([
      ...securitySchemes
        .filter(securityScheme => securityScheme.type === 'http')
        .map(securityScheme => {
          if (
            !SUPPORTED_HTTP_SCHEMES.includes(
              (securityScheme as OpenAPIV3.HttpSecurityScheme).scheme,
            )
          ) {
            throw new YError(
              'E_UNSUPPORTED_HTTP_SCHEME',
              (securityScheme as OpenAPIV3.HttpSecurityScheme).scheme,
            );
          }
          return (securityScheme as OpenAPIV3.HttpSecurityScheme).scheme;
        }),
      ...(hasOAuth ? ['bearer'] : []),
    ]),
  ];
  let hasBearerAuth = httpSchemes.includes('bearer');
  let hasAuthorizationApiKey = false;
  let hasAccessTokenApiKey = false;

  const securityParameters: OpenAPIV3.ParameterObject[] = securitySchemes
    .filter(securityScheme => securityScheme.type === 'apiKey')
    .map((securityScheme: OpenAPIV3.ApiKeySecurityScheme) => {
      if (securityScheme.in === 'cookie') {
        throw new YError(
          'E_UNSUPPORTED_API_KEY_SOURCE',
          'cookie',
          securityScheme.name,
        );
      }
      // This overlaps with OAuth/HTTP schemes
      if (
        securityScheme.in === 'header' &&
        securityScheme.name.toLowerCase() === 'authorization'
      ) {
        hasAuthorizationApiKey = true;
      }
      // This overlaps with OAuth and BearerAuth schemes
      if (
        securityScheme.in === 'query' &&
        securityScheme.name === 'access_token'
      ) {
        hasAccessTokenApiKey = true;
      }

      return {
        in: securityScheme.in,
        name:
          securityScheme.in === 'header'
            ? securityScheme.name.toLowerCase()
            : securityScheme.name,
        schema: {
          type: 'string',
        },
      } as OpenAPIV3.ParameterObject;
    })
    .concat(
      httpSchemes.length && !hasAuthorizationApiKey
        ? [
            {
              in: 'header',
              name: 'authorization',
              schema: {
                type: 'string',
                pattern: `(${httpSchemes
                  .map(
                    httpScheme =>
                      `(${
                        httpScheme[0]
                      }|${httpScheme[0].toUpperCase()})${httpScheme.slice(1)}`,
                  )
                  .join('|')}) .*`,
              },
            },
          ]
        : [],
    )
    .concat(
      hasBearerAuth && !hasAccessTokenApiKey
        ? [
            {
              in: 'query',
              name: 'access_token',
              schema: {
                type: 'string',
              },
            },
          ]
        : [],
    );
  return securityParameters;
}

export function prepareParametersValidators(
  ajv: Ajv,
  operationId: string,
  parameters: OpenAPIV3.ParameterObject[],
): { [name: string]: ajv.ValidateFunction } {
  return parameters.reduce((validators, parameter) => {
    if ('string' !== typeof parameter.name) {
      throw new YError('E_BAD_PARAMETER_NAME', operationId);
    }

    if (parameter.content) {
      throw new YError(
        'E_UNSUPPORTED_PARAMETER_DEFINITION',
        operationId,
        parameter.name,
        'content',
      );
    }

    if (parameter.style && 'simple' !== parameter.style) {
      throw new YError(
        'E_UNSUPPORTED_PARAMETER_DEFINITION',
        operationId,
        parameter.name,
        'style',
        parameter.style,
      );
    }

    if (!['query', 'header', 'path'].includes(parameter.in)) {
      throw new YError(
        'E_UNSUPPORTED_PARAMETER_DEFINITION',
        operationId,
        parameter.name,
        'in',
        parameter.in,
      );
    }

    let validator;

    try {
      validator = ajv.compile(parameter.schema);
    } catch (err) {
      throw YError.wrap(
        err,
        'E_BAD_PARAMETER_SCHEMA',
        operationId,
        parameter.name,
      );
    }

    validators[parameter.name] = _validateParameter.bind(
      null,
      parameter,
      validator,
    );
    return validators;
  }, {});
}

export function _validateParameter(
  parameter: OpenAPIV3.ParameterObject,
  validator: ajv.ValidateFunction,
  value: any,
) {
  if (parameter.required && 'undefined' === typeof value) {
    throw new HTTPError(
      400,
      'E_REQUIRED_PARAMETER',
      parameter.name,
      typeof value,
      value,
    );
  }
  if ('undefined' !== typeof value && !validator(value)) {
    throw new HTTPError(
      400,
      'E_BAD_PARAMETER',
      parameter.name,
      typeof value,
      value,
      validator.errors,
    );
  }
}

export function filterHeaders(
  parameters: OpenAPIV3.ParameterObject[],
  headers: { [name: string]: string },
): { [name: string]: string } {
  return (parameters || [])
    .filter(parameter => 'header' === parameter.in)
    .reduce((filteredHeaders, parameter) => {
      if (headers[parameter.name.toLowerCase()]) {
        filteredHeaders[camelCase(parameter.name)] =
          headers[parameter.name.toLowerCase()];
      }
      return filteredHeaders;
    }, {});
}
