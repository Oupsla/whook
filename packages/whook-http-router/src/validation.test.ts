import { extractParametersFromSecuritySchemes } from './validation';
import YError from 'yerror';

describe('extractParametersFromSecuritySchemes', () => {
  describe('should work', () => {
    it('with no security scheme', () => {
      expect(extractParametersFromSecuritySchemes([])).toMatchSnapshot();
    });

    it('with apiKey in query security scheme', () => {
      expect(
        extractParametersFromSecuritySchemes([
          {
            type: 'apiKey',
            in: 'query',
            name: 'yolo',
          },
        ]),
      ).toMatchSnapshot();
    });

    it('with apiKey in header security scheme', () => {
      expect(
        extractParametersFromSecuritySchemes([
          {
            type: 'apiKey',
            in: 'header',
            name: 'yolo',
          },
        ]),
      ).toMatchSnapshot();
    });

    it('with OAuth security scheme', () => {
      expect(
        extractParametersFromSecuritySchemes([
          {
            type: 'oauth2',
            flows: {},
          },
        ]),
      ).toMatchSnapshot();
    });

    it('with OpenId security scheme', () => {
      expect(
        extractParametersFromSecuritySchemes([
          {
            type: 'openIdConnect',
            openIdConnectUrl: 'http://test',
          },
        ]),
      ).toMatchSnapshot();
    });

    it('with header overlapping security schemes', () => {
      expect(
        extractParametersFromSecuritySchemes([
          {
            type: 'apiKey',
            in: 'header',
            name: 'Authorization',
          },
          {
            type: 'http',
            scheme: 'bearer',
          },
        ]),
      ).toMatchSnapshot();
    });

    it('with query overlapping security schemes', () => {
      expect(
        extractParametersFromSecuritySchemes([
          {
            type: 'apiKey',
            in: 'query',
            name: 'access_token',
          },
          {
            type: 'oauth2',
            flows: {},
          },
        ]),
      ).toMatchSnapshot();
    });

    it('with nested security scheme', () => {
      expect(
        extractParametersFromSecuritySchemes([
          {
            type: 'apiKey',
            in: 'header',
            name: 'Authorization',
          },
          {
            type: 'apiKey',
            in: 'header',
            name: 'yolo',
          },
          {
            type: 'apiKey',
            in: 'query',
            name: 'yolo',
          },
          {
            type: 'apiKey',
            in: 'query',
            name: 'access_token',
          },
          {
            type: 'http',
            scheme: 'bearer',
          },
          {
            type: 'oauth2',
            flows: {},
          },
          {
            type: 'openIdConnect',
            openIdConnectUrl: 'http://test',
          },
        ]),
      ).toMatchSnapshot();
    });
  });
  describe('should fail', () => {
    it('with unsupported security scheme', () => {
      try {
        extractParametersFromSecuritySchemes([
          {
            type: 'http',
            scheme: 'mutual',
          },
        ]);
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
        }).toMatchSnapshot();
      }
    });

    it('with unsupported API scheme source', () => {
      try {
        extractParametersFromSecuritySchemes([
          {
            type: 'apiKey',
            in: 'cookie',
            name: 'access_token',
          },
        ]);
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
        }).toMatchSnapshot();
      }
    });
  });
});
