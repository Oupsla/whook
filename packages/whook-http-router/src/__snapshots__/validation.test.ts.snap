// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractParametersFromSecuritySchemes should fail with unsupported API scheme source 1`] = `
Object {
  "errorCode": "E_UNSUPPORTED_API_KEY_SOURCE",
  "errorParams": Array [
    "cookie",
    "access_token",
  ],
}
`;

exports[`extractParametersFromSecuritySchemes should fail with unsupported security scheme 1`] = `
Object {
  "errorCode": "E_UNSUPPORTED_HTTP_SCHEME",
  "errorParams": Array [
    "mutual",
  ],
}
`;

exports[`extractParametersFromSecuritySchemes should work with OAuth security scheme 1`] = `
Array [
  Object {
    "in": "header",
    "name": "authorization",
    "schema": Object {
      "pattern": "((b|B)earer) .*",
      "type": "string",
    },
  },
  Object {
    "in": "query",
    "name": "access_token",
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`extractParametersFromSecuritySchemes should work with OpenId security scheme 1`] = `
Array [
  Object {
    "in": "header",
    "name": "authorization",
    "schema": Object {
      "pattern": "((b|B)earer) .*",
      "type": "string",
    },
  },
  Object {
    "in": "query",
    "name": "access_token",
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`extractParametersFromSecuritySchemes should work with apiKey in header security scheme 1`] = `
Array [
  Object {
    "in": "header",
    "name": "yolo",
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`extractParametersFromSecuritySchemes should work with apiKey in query security scheme 1`] = `
Array [
  Object {
    "in": "query",
    "name": "yolo",
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`extractParametersFromSecuritySchemes should work with header overlapping security schemes 1`] = `
Array [
  Object {
    "in": "header",
    "name": "authorization",
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "in": "query",
    "name": "access_token",
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`extractParametersFromSecuritySchemes should work with nested security scheme 1`] = `
Array [
  Object {
    "in": "header",
    "name": "authorization",
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "in": "header",
    "name": "yolo",
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "in": "query",
    "name": "yolo",
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "in": "query",
    "name": "access_token",
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`extractParametersFromSecuritySchemes should work with no security scheme 1`] = `Array []`;

exports[`extractParametersFromSecuritySchemes should work with query overlapping security schemes 1`] = `
Array [
  Object {
    "in": "query",
    "name": "access_token",
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "in": "header",
    "name": "authorization",
    "schema": Object {
      "pattern": "((b|B)earer) .*",
      "type": "string",
    },
  },
]
`;
